File: ts/src/sdto/FluentAction.ts
import { FieldSpec } from "./FieldSpec";
import { buildSAE } from "../sae";

// FluentAction is the fluent builder for constructing validated actions
export class FluentAction {
  private actionType: string;
  private schema: Record<string, FieldSpec>;
  private data: Record<string, unknown> = {};
  private errors: Error[] = [];

  constructor(actionType: string, schema: Record<string, FieldSpec>) {
    this.actionType = actionType;
    this.schema = schema;
  }

  // Set validates and stores a field value
  set(key: string, value: unknown): this {
    const spec = this.schema[key];
    if (!spec) {
      this.errors.push(new Error(`unknown field: ${key}`));
      return this;
    }

    const err = validateValue(value, spec);
    if (err) {
      this.errors.push(new Error(`field ${key}: ${err.message}`));
      return this;
    }

    this.data[key] = value;
    return this;
  }

  // Finalize returns SAE bytes or throws aggregated errors
  // Matches Go's Finalize() which returns ([]byte, error)
  finalize(): Buffer {
    // Check for missing required fields (all schema fields are required)
    for (const key of Object.keys(this.schema)) {
      if (!(key in this.data)) {
        this.errors.push(new Error(`missing required field: ${key}`));
      }
    }

    if (this.errors.length > 0) {
      const messages = this.errors.map((e) => e.message).join("; ");
      throw new Error(messages);
    }

    // Build SAE using JCS canonicalizer (matches Go's sae.BuildSAE)
    return buildSAE(this.actionType, this.data);
  }
}

// Validation helper functions (moved outside class to match Go pattern)

function validateValue(value: unknown, spec: FieldSpec): Error | null {
  switch (spec.type) {
    case "string":
      return validateString(value, spec);
    case "number":
      return validateNumber(value, spec);
    case "sign":
      return validateSign(value, spec);
    default:
      return new Error(`unknown type "${spec.type}"`);
  }
}

function validateSign(value: unknown, spec: FieldSpec): Error | null {
  // Sign value must be string (type is defined in schema)
  if (typeof value !== "string") {
    return new Error("sign field expects string value");
  }

  // Can extend: validate format based on spec.enum[0] (hex/base64 etc)
  if (value.length === 0) {
    return new Error("sign value cannot be empty");
  }

  return null;
}

function validateString(value: unknown, spec: FieldSpec): Error | null {
  if (typeof value !== "string") {
    return new Error("expected string");
  }

  // enum check
  if (spec.enum && spec.enum.length > 0) {
    if (!spec.enum.includes(value)) {
      return new Error(`value "${value}" not in enum`);
    }
    return null;
  }

  // length boundary (numeric parsing)
  if (spec.min !== undefined) {
    const minLen = parseInt(spec.min, 10);
    if (!isNaN(minLen) && value.length < minLen) {
      return new Error(`string length ${value.length} < min ${minLen}`);
    }
  }
  if (spec.max !== undefined) {
    const maxLen = parseInt(spec.max, 10);
    if (!isNaN(maxLen) && value.length > maxLen) {
      return new Error(`string length ${value.length} > max ${maxLen}`);
    }
  }

  return null;
}

function validateNumber(value: unknown, spec: FieldSpec): Error | null {
  if (typeof value !== "number") {
    return new Error("expected number");
  }

  if (spec.min !== undefined) {
    const minVal = parseFloat(spec.min);
    if (!isNaN(minVal) && value < minVal) {
      return new Error("number < min");
    }
  }
  if (spec.max !== undefined) {
    const maxVal = parseFloat(spec.max);
    if (!isNaN(maxVal) && value > maxVal) {
      return new Error("number > max");
    }
  }

  return null;
}

// ValidateData validates a map against schema (for server-side verification)
// Matches Go's ValidateData function
export function validateData(
  data: Record<string, unknown>,
  schema: Record<string, FieldSpec>
): void {
  const errors: Error[] = [];

  // Check all required fields in schema exist
  for (const [key, spec] of Object.entries(schema)) {
    const value = data[key];
    if (value === undefined) {
      errors.push(new Error(`missing field: ${key}`));
      continue;
    }
    const err = validateValue(value, spec);
    if (err) {
      errors.push(new Error(`field ${key}: ${err.message}`));
    }
  }

  // Check no extra fields
  for (const key of Object.keys(data)) {
    if (!(key in schema)) {
      errors.push(new Error(`unknown field: ${key}`));
    }
  }

  if (errors.length > 0) {
    throw new Error(errors.map((e) => e.message).join("; "));
  }
}

// Factory function to match Go's NewAction
export function newAction(
  actionType: string,
  schema: Record<string, FieldSpec>
): FluentAction {
  return new FluentAction(actionType, schema);
}

File: go/pkg/vax/sdto/FluentAction.go
package sdto

import (
	"errors"
	"fmt"
	"math/big"
	"strconv"

	"vax/pkg/vax/sae"
)

// FluentAction 是你給 Consumer 的「量尺」
type FluentAction struct {
	actionType string
	schema     map[string]FieldSpec // 從後端拉回來的驗證規則
	data       map[string]any
	errs       []error
}

func NewAction(actionType string, rules map[string]FieldSpec) *FluentAction {
	return &FluentAction{
		actionType: actionType,
		schema:     rules,
		data:       make(map[string]any),
	}
}

// Set 在賦值的瞬間進行驗證
func (f *FluentAction) Set(key string, value any) *FluentAction {
	spec, exists := f.schema[key]
	if !exists {
		f.errs = append(f.errs, fmt.Errorf("unknown field: %s", key))
		return f
	}

	if err := validateValue(value, spec); err != nil {
		f.errs = append(f.errs, fmt.Errorf("field %s: %w", key, err))
		return f
	}

	f.data[key] = value
	return f
}

func validateValue(value any, c FieldSpec) error {
	switch c.Type {
	case "string":
		return validateString(value, c)
	case "number":
		return validateNumber(value, c)
	case "sign":
		return validateSign(value, c)
	default:
		return fmt.Errorf("unknown type %q", c.Type)
	}
}

func validateSign(value any, c FieldSpec) error {
	// 簽名值只能是 string（類型已在 schema 層定義）
	v, ok := value.(string)
	if !ok {
		return errors.New("sign field expects string value")
	}

	// 可擴展：根據 c.Enum[0] 做格式驗證（hex/base64 等）
	if len(v) == 0 {
		return errors.New("sign value cannot be empty")
	}

	return nil
}

func validateString(value any, c FieldSpec) error {
	v, ok := value.(string)
	if !ok {
		return errors.New("expected string")
	}

	// enum
	if len(c.Enum) > 0 {
		for _, allowed := range c.Enum {
			if v == allowed {
				return nil
			}
		}
		return fmt.Errorf("value %q not in enum", v)
	}

	// length boundary (數值解析)
	if c.Min != nil {
		minLen, err := strconv.Atoi(*c.Min)
		if err == nil && len(v) < minLen {
			return fmt.Errorf("string length %d < min %d", len(v), minLen)
		}
	}
	if c.Max != nil {
		maxLen, err := strconv.Atoi(*c.Max)
		if err == nil && len(v) > maxLen {
			return fmt.Errorf("string length %d > max %d", len(v), maxLen)
		}
	}

	return nil
}

func validateNumber(value any, c FieldSpec) error {
	var v float64

	switch n := value.(type) {
	case int:
		v = float64(n)
	case int64:
		v = float64(n)
	case float32:
		v = float64(n)
	case float64:
		v = n
	default:
		return errors.New("expected number")
	}

	if c.Min != nil {
		if !compareNumber(v, *c.Min, ">=") {
			return fmt.Errorf("number < min")
		}
	}
	if c.Max != nil {
		if !compareNumber(v, *c.Max, "<=") {
			return fmt.Errorf("number > max")
		}
	}

	return nil
}

func compareNumber(value float64, bound string, op string) bool {
	v := new(big.Rat).SetFloat64(value)

	b := new(big.Rat)
	if _, ok := b.SetString(bound); !ok {
		return false
	}

	switch op {
	case ">=":
		return v.Cmp(b) >= 0
	case "<=":
		return v.Cmp(b) <= 0
	default:
		return false
	}
}

// Finalize 最終產出 SAE
func (f *FluentAction) Finalize() ([]byte, error) {
	// Check for missing required fields (all schema fields are required)
	for key := range f.schema {
		if _, exists := f.data[key]; !exists {
			f.errs = append(f.errs, fmt.Errorf("missing required field: %s", key))
		}
	}

	if len(f.errs) > 0 {
		// Aggregate errors into a single message
		msg := ""
		for i, err := range f.errs {
			if i > 0 {
				msg += "; "
			}
			msg += err.Error()
		}
		return nil, errors.New(msg)
	}
	// 調用你剛剛寫好的 SAE.BuildSAE
	return sae.BuildSAE(f.actionType, f.data)
}

// ValidateData validates a map against schema (for server-side verification)
func ValidateData(data map[string]any, schema map[string]FieldSpec) error {
	var errs []error

	// Check all required fields in schema exist
	for key, spec := range schema {
		value, exists := data[key]
		if !exists {
			errs = append(errs, fmt.Errorf("missing field: %s", key))
			continue
		}
		if err := validateValue(value, spec); err != nil {
			errs = append(errs, fmt.Errorf("field %s: %w", key, err))
		}
	}

	// Check no extra fields
	for key := range data {
		if _, exists := schema[key]; !exists {
			errs = append(errs, fmt.Errorf("unknown field: %s", key))
		}
	}

	if len(errs) > 0 {
		// Aggregate errors into a single message
		msg := ""
		for i, err := range errs {
			if i > 0 {
				msg += "; "
			}
			msg += err.Error()
		}
		return errors.New(msg)
	}
	return nil
}
